doc/memory_statistics/memory_statistics_enhancement.md

```diff
@@ -0,0 +1,338 @@
+
+## Table of Contents
+
+- [Scope](#scope)
+- [Definitions/Abbreviations](#definitionsabbreviations)
+- [Overview](#overview)
+- [Functional Requirements](#functional-requirements)
+- [Architecture Design](#architecture-design)
+- [High-Level Design](#high-level-design)
+  - [Core Functionalities](#core-functionalities)
+  - [Sequence Diagrams](#sequence-diagrams)
+- [SAI API](#sai-api)
+- [Configuration and Management](#configuration-and-management)
+  - [CLI/YANG Model Enhancements](#cliyang-model-enhancements)
+    - [CLI Commands](#cli-commands)
+      - [Config Commands](#config-commands)
+      - [Show Commands](#show-commands)
+  - [Daemon Configuration Management](#daemon-configuration-management)
+  - [Config DB Enhancements](#config-db-enhancements)
+- [Warmboot and Fastboot Design Impact](#warmboot-and-fastboot-design-impact)
+- [Testing Requirements/Design](#testing-requirementsdesign)
+  - [Unit Test Cases](#unit-test-cases)
+  - [System Test Cases](#system-test-cases)
+- [Future Work](#future-work)
+
+---
+
+## Scope
+
+This High-Level Design (HLD) document enhances the existing Memory Statistics feature in SONiC by extending memory metrics to include Docker and process memory, adding memory leak detection, and enabling remote log access via gNMI.
+
+---
+
+## Definitions/Abbreviations
+
+| Sr No. | Term         | Definition  |
+|--------|-------------|------------|
+| 1      | gNMI        | Network Management Interface, a protocol for network management and telemetry |
+| 2      | gRPC        | Remote Procedure Call, a high-performance framework for communication between network components |
+| 3      | Telemetry   | A mechanism to collect and export real-time operational data from the system |
+| 4      | Log Processing | The method of formatting and organizing log files for efficient retrieval and analysis |
+
+---
+
+## Overview
+This High-Level Design (HLD) enhances SONiC’s Memory Statistics feature, originally limited to system-level memory monitoring via CLI, into a comprehensive solution that now includes Docker containers and individual processes memory metrics. It introduces memory leak detection to prevent resource exhaustion and integrates gNMI for remote log access, reducing local dependency. This upgrade provides complete visibility across memory metrics with user-defined sampling and retention periods. Moreover, it also helps with early memory leak detection and efficient remote retrieval.
+
+---
+
+## Functional Requirements
+This section outlines the functional requirements necessary for implementing this HLD in SONiC:
+
+- **Monitoring Capabilities:** The system must monitor memory metrics for system (Total, Used, Free, Available, Cached, Buffers, Shared), Docker containers and individual processes.
+- **Memory Leak Detection:** The feature must analyze memory usage trends over time to detect potential leaks and report them via CLI.
```

Comment:
What is the practical benefit of storing memory histograms on the switch itself? Would it be more efficient to pull this data externally and analyze it off-box?